File: mysite/simulation/static/js/sim.js
const BUFFER = 9.0;
const TRUE_NORTH = 0; //Robot.random_angle();

class Simulation extends Robot {

	static random_state(width, length, buffer) {
		var theta = Robot.random_angle();

		var fl = Robot.rotate(Robot.FRONT_LEFT, theta);
		var fr = Robot.rotate(Robot.FRONT_RIGHT, theta);
		var bl = Robot.rotate(Robot.BACK_LEFT, theta);
		var br = Robot.rotate(Robot.BACK_RIGHT, theta);
		var robot_x = [fl.x, fr.x, bl.x, br.x];
		var robot_y = [fl.y, fr.y, bl.y, br.y];

		var min_x = Math.min(...robot_x);
		var max_x = Math.max(...robot_x);
		var min_y = Math.min(...robot_y);
		var max_y = Math.max(...robot_y);

		var state = new State(
			Robot.random_float(buffer - min_x, length - max_x - buffer),
			Robot.random_float(buffer - min_y, width - max_y - buffer),
			theta,
			0
		);

		return state;
	}


	static add_noise(input, ratio) {
		return input + Robot.gaussian() * input * ratio + Robot.gaussian() * ratio;
	}


	constructor() {
		super(Simulation.random_state(sandbox.width_cm, sandbox.length_cm, BUFFER));
		this.period = 5.0;
		this.active = true;	
	}


	state_simulator(left, right) {
		
		for (var step = 0; step < left.length; step += 1) {

			let left_ratio = left[step] / Robot.ANALOG_MAX;
			let right_ratio = right[step] / Robot.ANALOG_MAX;

			let ratio_diff = right_ratio - left_ratio;
			let ratio_sum = right_ratio + left_ratio;

			let max_dist = Robot.MAX_CM_PER_MS * this.period;
			let dist = ratio_sum * max_dist / 2;
			let turn_radius = Robot.AXLE_LENGTH * ratio_sum / (2 * ratio_diff);
			let dtheta = ratio_diff * max_dist / Robot.AXLE_LENGTH;

			let orient = Robot.next_orientation(this.state, dist, turn_radius, dtheta);
			this.prev.state = this.state;
			this.state.update(orient.x, orient.y, orient.theta, dist);
		}
	}


	get_front_dist() {
		this.prev.front_laser.dist = this.front_laser.dist;
		this.front_laser.dist = Robot.distance_from_point(
			new State(this.front_laser.pos.x,
								this.front_laser.pos.y,
								this.state.theta),
			sandbox).dist;

		return this.front_laser.dist;
	}


	get_right_dist() {
		this.prev.right_laser.dist = this.right_laser.dist;
		this.right_laser.dist = Robot.distance_from_point(
			new State(this.right_laser.pos.x,
								this.right_laser.pos.y,
								Robot.reduce_angle(this.state.theta - Math.PI/2)),
			sandbox).dist;

		return this.right_laser.dist;
	}



	get_mag() {
		this.prev.imu.mag = this.imu.mag;
		this.imu.mag = Robot.reduce_angle(this.state.theta - TRUE_NORTH);
		return this.imu.mag;
	}


	get_gyro() {
		this.prev.imu.gyro = this.imu.gyro;
		this.imu.gyro = Robot.reduce_angle(this.imu.mag - this.prev.imu.mag) * 
										this.period / 1000;
		return this.imu.gyro;
	}


	get_accel() {
		this.prev.imu.accel = this.imu.accel;
		this.imu.accel = (this.state.v - this.prev.state.v) * 
											this.period / 1000;
		return this.imu.accel;
	}


	update_sensors() {
		return {
			front : [Simulation.add_noise(this.get_front_dist(), 0.05)],
			right : [Simulation.add_noise(this.get_right_dist(), 0.05)],
			accel : [Simulation.add_noise(this.get_accel(), 0.05)],
			mag   : [Simulation.add_noise(this.get_mag(), 0.05)],
			gyro  : [Simulation.add_noise(this.get_gyro(), 0.05)],
		};
	}
}


function simulator(simbot) {

	if (simbot.active) {

		// detect new inputs
		var target = http_IF.receive_target();
		var controls = http_IF.receive_controls();

		if (target != null) {
			http_IF.ajax = true;
			console.log("target exists");
			// UNIMPLEMENTED

		// read a list of inputs and react
		} else if (controls != null) {

			controls = controls.slice(5,-1);
			let count = 0;
			let left = [];
			let right = [];

			while (count < controls.length) {

				left.push(parseInt(controls.substr(count, 4)));
				right.push(parseInt(controls.substr(count + 4, 4)));
				count += 8;

			}

			simbot.state_simulator(left, right);
		}

		// update relative parts positions
		simbot.update_positions();

		// update sensors
		http_IF.send_sensor_data(simbot.update_sensors());

		// draw
		draw_robot(simbot.state, false);

	}

	if (fend_IF.receive_options("simulate")) {
		simbot.active = false;
		fend_IF.simululation_off();
	}
}

File: mysite/simulation/alt/sim.py
import asyncio
import math
import random
from math import pi
from robot import Robot, Position, State, Box
from interface import HTTP_Interface

Sandbox = Box(80.0, 100.0)

BUFFER = 9.0
TRUE_NORTH = round(random.uniform(-pi, pi), 2)

class SimulatedRobot(Robot):

	# Generates random legal starting state
	def random_state(length, width, buffer):
		theta = random.uniform(-pi, pi)
		
		# robot extremities in robot frame
		fl = Robot.rotate(Robot.FRONT_LEFT, theta)
		fr = Robot.rotate(Robot.FRONT_RIGHT, theta)
		bl = Robot.rotate(Robot.BACK_LEFT, theta)
		br = Robot.rotate(Robot.BACK_RIGHT, theta)
		robot_x = [fl.x, fr.x, bl.x, br.x]
		robot_y = [fl.y, fr.y, bl.y, br.y]

		# robot extremities in field frame
		min_x = min(robot_x)
		max_x = max(robot_x)
		min_y = min(robot_y)
		max_y = max(robot_y)

		return State(
			random.uniform(buffer - min_x, length - max_x - buffer),
			random.uniform(buffer - min_y, width - max_y - buffer),
			theta
		)


	def add_noise(input, ratio):
		return input + random.gauss(0, 1) * input * ratio


	def __init__(self):
		super().__init__(SimulatedRobot.random_state(Sandbox.length, Sandbox.width, BUFFER))
		self._period = 5.0
		self._active = True


	def state_simulator(self, left, right):
		# input sequence
		left_ratio = left / Robot.ANALOG_MAX
		right_ratio = right / Robot.ANALOG_MAX

		for step in range(len(left)):

			ratio_diff = right_ratio[step] - left_ratio[step]

			max_dist = Robot.MAX_CM_PER_MS * self._period

			avg_ratio = (left_ratio + right_ratio) / 2;

			dist = avg_ratio * max_dist;

			turn_radius = left_ratio[step] * Robot.AXLE_LENGTH / ratio_diff

			d_theta = ratio_diff * max_dist / Robot.AXLE_LENGTH

			# set robot as axis of rotation to find position of center
			if abs(turn_radius) < Sandbox.width / 2:
				rel_pos = Position( -(turn_radius + Robot.AXLE_LENGTH / 2), 0)
				center = Robot.transform(rel_pos, self._state)

				# travel along circumference of circle
				rel_pos = Position(self._state.x - center.x,
													self._state.y - center.y)
				d_state = State(center.x, center.y, d_theta)

			# straight path
			else:

				# set original state as axis of rotation
				rel_pos = Position(0, avg_dist)
				d_state = self._state
				
			# update state
			cur_pos = Robot.transform(rel_pos, d_state)
			self._prev._state = self._state
			self._state = State(cur_pos.x, 
													cur_pos.y, 
													self._state.theta + d_theta,
													avg_dist)



	def get_front_dist(self):
		self._prev._front_laser['dist'] = self._front_laser['dist']
		self._front_laser['dist'] = Robot.distance_from_point(
				State(self._front_laser['pos'].x,
							self._front_laser['pos'].y,
							self._state.theta))[0]
		return self._front_laser['dist']

		
	def get_right_dist(self):
		self._prev._right_laser['dist'] = self._right_laser['dist']
		self._right_laser['dist'] = Robot.distance_from_point(
				State(self._right_laser['pos'].x,
							self._right_laser['pos'].y, 
							self._state.theta - pi/2))[0]
		return self._right_laser['dist']


	def get_mag(self):
		self._prev._imu['mag'] = self._imu['mag']
		self._imu['mag'] = self._state.theta - TRUE_NORTH
		return self._imu['mag']


	def get_gyro(self): # scaled to cm/s
		self._prev._imu['gyro'] = self._imu['gyro']
		self._imu['gyro'] = (self.imu['mag'] - self._prev.imu['mag']) * \
			1000 / self._period
		return self._imu['gyro']


	def get_accel(self): # scaled to cm/s
		self._prev._imu['accel'] = self._imu['accel']
		self._imu['accel'] = (self._state.vel - self._prev._state.vel) * \
			1000 / self._period
		return self._imu['accel']


	# Returns a dictionary of sensors
	def update_sensors(self):
		return {
			"front": SimulatedRobot.add_noise(self.get_front_dist(), 0.05),
			"right": SimulatedRobot.add_noise(self.get_right_dist(), 0.05),
			"mag": SimulatedRobot.add_noise(self.get_mag(), 0.05),
			"gyro": SimulatedRobot.add_noise(self.get_gyro(), 0.05),
			"accel": SimulatedRobot.add_noise(self.get_accel(), 0.05)
		}


async def simulate_loop(sim, loop=None):

	while True:
		# break if turned off
		if sim._active == False:

			# undraw
			
			break
		print("what")
		# check for inputs
		ctrl_str = HTTP_Interface.receive_controls()
		controls = ctrl_str[6:].split(",")
		left = controls[0::2]
		right = controls[1::2]

		# iterate the state
		sim.state_simulator(left, right)

		# recalculate simulated robot positions
		sim.update_positions()

		# simulate HTTP response to send sensor data
		HTTP_Interface.send_sensor_data(sim.update_sensors())

		# draw

		# let other tasks run
		await asyncio.sleep(sim._period/1000.0)


def schedule_simulator(loop=None):
	print("bye")
	# create simulation
	sim = SimulatedRobot()
	if loop is None:
		loop = asyncio.get_event_loop()
	sim._loop = loop
	print("hi")
	# schedule task loop
	sim._task = asyncio.ensure_future(simulate_loop(sim, loop))
	return sim

def stop_simulator(sim, loop):
	sim._active = False
	sim._task.cancel()
